stm32-modules/common/module-startup/module-startup.S

```diff
@@ -0,0 +1,186 @@
+/**
+  ******************************************************************************
+  * @file   module-startup.S
+  * @brief  Opentrons STM32 Modules startup application to provide a safety net
+  *         in the case of device update failure.
+  *         This application:
+  *           - Provides an (incomplete) vector table with handlers for faults
+  *             which will jump to the STM32 internal bootloader
+  *           - Jumps to the second flash sector on the system in order to
+  *             start the actual module firmware. If no firmware is installed,
+  *             a hardfault should bring the device back to the DFU 
+  *             bootloader.
+  ******************************************************************************
+  */
+
+  .syntax unified
+	.cpu cortex-m4
+	.fpu softvfp
+	.thumb
+
+.global	g_pfnVectors
+.global Fault_Handler
+.global Post_Fault_Handler
+
+.equ  BootRAM,        0xF1E0F85F
+
+/**
+ * @brief  This is the code that gets called when the processor first
+ *          starts execution following a reset event. This simply jumps
+ *          to the actual application on the second page.
+ * @param  None
+ * @retval : None
+*/
+
+    .section	.text.Reset_Handler
+	.weak	Reset_Handler
+	.type	Reset_Handler, %function
+Reset_Handler:
+    ldr     r0, =_estack 
+    mov     sp, r0 
+    /* Jump to the user application */
+    ldr     r0, =_user_app_start
+    ldr     r1, [r0]
+    /* As an easy way to check for a missing program, check if the reset
+     * vector we're expecting points to 0xFFFFFFFx. If that's the case,
+     * there is definitely not a valid program to run. */
+    ldr     r2, =0xFFFFFFF0
+    and     r1, r1, r2
+    cmp     r1, r2 
+    bcs Post_Fault_Handler /* jump if upper 7 nibbles are all 0xF */
```

Comment:
Your logic is 100% right, but this works because the Carry Flag is inverted for `subs` operations. https://developer.arm.com/documentation/dui0801/a/CIADCDHH

Reply:
Oh how I love digital logic